### PR TITLE
Refactor security plugin to provide authorize functions

### LINKED ISSUE #7: No security attribute on method generates invalid code.
https://github.com/goadesign/plugins/blob/6127466225ad03c6db8b215094b49a72d09f9933/security/generate.go#L169

If a method has no DSL Security Attribute example Security(....)  then it fails on build of generated project with "undefined: NewMethodName"  

Changing the above line to this 
```
{{ .FieldName }}: {{ .VarName }}{{ if not .Requirements}}Endpoint{{ end }}{{ if .Requirements }}({{ .NonSecureVarName }}{{ end }}(s){{ if .Requirements }}){{ end }},```
```

It seems to fix the issue, compiles and  method is *seems* to execute correctly.  However there is probably easier way to fix this.

### PR DESCRIPTION
**Pull request by nitinmohan87**
* Users can provide custom auth functions for a scheme type
* Fixes #7 
* Added security definitions to openapi

### CODE REVIEW

**Comment by raphael** [comment]
Whoops - lost track of this one, what's the status? should it be merged?

File: security/examples/calc/adder/adder.go [hunk 1]
```diff
@@ -30,36 +22,11 @@ func NewAdder(logger *log.Logger) adder.Service {
...
-	// you shouldn't do that in production...
-	s.logger.Printf("Add request authorized with token '%s'", t.Raw)
-
 	return p.A + p.B, nil
 }
+
+// AuthAPIKeyFn implements the authorization logic for APIKey scheme.
+func AuthAPIKeyFn(ctx context.Context, key string, s *security.APIKeyScheme) (context.Context, error) {
+	// Add authorization logic
+	return ctx, nil
```
[the remark this answers is not in the document]
Comment (nitinmohan87) [reply security/examples/calc/adder/adder.go]: How about just checking if the key is not empty?

File: security/design/security.go [hunk 2]
```diff
@@ -172,7 +172,7 @@ func (s *SchemeExpr) Type() string {
 	case JWTKind:
 		return "JWT"
 	default:
-		return "[unknown]"
+		panic(fmt.Sprintf("unknown scheme kind: %#v", s.Kind)) // bug
```
Comment (raphael) [inline security/design/security.go]: I think there are places in the code where we check whether the type is `[unknown]` - these checks are not needed anymore either correct?

File: security/examples/calc/adder/adder.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (raphael) [reply security/examples/calc/adder/adder.go]: that would work too. Using a simple constant is not much more work and is a bit more realistic though :)

**Comment by raphael** [comment]
Thank you!

**Pull request merged**